AS-368: Add Request-Id to every log message associated with an API call

MatthewBemis opened the pull request:
Log messages will look something like:

```
2020-06-19 09:40:41.535 pn3N7J3M ERROR 17147 --- [nio-8080-exec-1] b.t.w.a.c.GlobalExceptionHandler         :    cause: bio.terra.workspace.common.exception.DuplicateWorkspaceException@2696eb2[causes=<null>,statusCode=400 BAD_REQUEST]
```

This does _not_ add the ID to anything that happens asynchronously within stairway. It will be added to the message that says the flight was launched, but no subsequent messages from within the flight.

MatthewBemis commented on src/main/java/bio/terra/workspace/mdc/MdcLogEnhancerFilter.java:
```diff
@@ -0,0 +1,35 @@
+    HttpServletRequest httpRequest = (HttpServletRequest) servletRequest;
+
+    String requestId =
+        ((requestId = httpRequest.getHeader("Request-Id")) != null)
```
`getHeader` is case-insensitive, fortunately

davidangb replied:
could this be `X-Request-ID` or `X-Correlation-ID`, maybe even look for both before generating one ourselves? https://en.wikipedia.org/wiki/List_of_HTTP_header_fields

MatthewBemis commented on src/main/java/bio/terra/workspace/mdc/MdcLogEnhancerFilter.java:
```diff
@@ -0,0 +1,35 @@
+            ? requestId
+            : generateRequestId();
+
+    MDC.put("requestId", requestId + " ");
```
I added a space after the `requestId` as extra padding because when testing locally it was bumping up against the log-level in the message i.e. it was producing messages that looked like `pn3N7J3MERROR`. That may have just been formatting in my terminal not properly showing a tab or something. If there are objects I can remove.

davidangb commented on src/main/resources/application.properties:
```diff
@@ -20,3 +20,5 @@ sam.basePath=${SAM_ADDRESS}
 datarepo.instances.terra-dev=https://jade.datarepo-dev.broadinstitute.org
 datarepo.instances.terra-prod=https://jade-terra.datarepo-prod.broadinstitute.org
+
+logging.pattern.level = %X{requestId}%5p
```
I think if you change this to `%X{requestId} %5p` (note the space) then you won't have to append a space to the requestid in `MdcLogEnhancerFilter`. See https://docs.spring.io/autorepo/docs/spring-boot/1.4.0.RELEASE/reference/html/boot-features-logging.html (search for "You can add MDC and other ad-hoc content")

davidangb commented on src/main/java/bio/terra/workspace/mdc/MdcLogEnhancerFilter.java:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  private String generateRequestId() {
+    Hashids hashids = new Hashids("requestIdSalt", 8); // min length will be 8
```
is it worth creating the hashids instance at the class level? That would avoid calling the constructor for every request. The call to `encode` *looks* threadsafe.

MatthewBemis commented:
@davidangb I've addressed your comments. +1 for creating a followup story for stairway transactions. I'll create that.

I also added another filter to add `X-Request-ID` to the HTTP response headers for each request. Seems like it could be useful. Let me know if you think it's superfluous!
